src/pycapsule.rs

```diff
@@ -0,0 +1,330 @@
+    destructor(value)
+}
+
+impl PyCapsule {
+    /// Constructs a new capsule of whose contents are `T` associated with `name`.
+    pub fn new<'py, T: 'static>(py: Python<'py>, value: T, name: &CStr) -> PyResult<&'py Self> {
+        Self::new_with_destructor(py, value, name, std::mem::drop)
+    }
+
+    /// Constructs a new capsule of whose contents are `T` associated with `name`
+    /// Provide a destructor for when `PyCapsule` is destroyed it will be passed the capsule.
+    pub fn new_with_destructor<'py, T: 'static, F: FnOnce(T)>(
+        py: Python<'py>,
+        value: T,
+        name: &CStr,
+        destructor: F,
+    ) -> PyResult<&'py Self> {
+        let val = Box::new(CapsuleContents::new(value, destructor));
+
+        let cap_ptr = unsafe {
+            ffi::PyCapsule_New(
+                Box::into_raw(val) as *mut c_void,
+                name.as_ptr(),
+                Some(capsule_destructor::<T, F>),
+            )
+        };
+        if cap_ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(unsafe { py.from_owned_ptr::<PyCapsule>(cap_ptr) })
+        }
+    }
+
+    /// Import an existing capsule.
+    ///
+    /// The `name` should match the path to module attribute exactly in the form
+    /// of `module.attribute`, which should be the same as the name within the
+    /// capsule. `no_block` indicates to use
+    /// [PyImport_ImportModuleNoBlock()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModuleNoBlock)
+    /// or [PyImport_ImportModule()](https://docs.python.org/3/c-api/import.html#c.PyImport_ImportModule)
+    /// when accessing the capsule.
+    ///
+    /// ## Safety
+    /// This is unsafe, as there is no guarantee when casting `*mut void` into `T`.
+    pub unsafe fn import<'py, T>(py: Python<'py>, name: &CStr, no_block: bool) -> PyResult<&'py T> {
+        let ptr = ffi::PyCapsule_Import(name.as_ptr(), no_block as c_int);
+        if ptr.is_null() {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(&*(ptr as *const T))
+        }
+    }
+
+    /// Set a context pointer in the capsule to `T`
+    pub fn set_context<'py, T: 'static>(&self, py: Python<'py>, context: T) -> PyResult<()> {
+        let ctx = Box::new(context);
+        let result =
+            unsafe { ffi::PyCapsule_SetContext(self.as_ptr(), Box::into_raw(ctx) as _) as u8 };
+        if result != 0 {
+            Err(PyErr::fetch(py))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get a reference to the context `T` in the capsule, if any.
+    ///
+    /// ## Safety
+    ///
+    /// This is unsafe, as there is no guarantee when casting `*mut void` into `T`.
+    pub unsafe fn get_context<T>(&self, py: Python) -> PyResult<Option<&T>> {
+        let ctx = ffi::PyCapsule_GetContext(self.as_ptr());
+        if ctx.is_null() {
+            if self.is_valid() & PyErr::occurred(py) {
```

Comment:
Bitwise operator looks potentially unintentional

```suggestion
            if self.is_valid() && PyErr::occurred(py) {
```